pkg/cache/file_cache.go

```diff
@@ -54,11 +55,15 @@ func (c *FileCache) HasLayer(layer types.BlobInfo) bool {
 func (c *FileCache) SetLayer(layer types.BlobInfo, r io.Reader) (io.ReadCloser, error) {
 	layerId := layer.Digest.String()
 	fullpath := filepath.Join(c.RootDir, layerId)
-	entry, err := os.Create(fullpath)
+	// Write the entry atomically. First write it to a .tmp name, then rename to the correct one.
+	f, err := ioutil.TempFile("", "")
 	if err != nil {
 		return nil, err
 	}
-	if _, err := io.Copy(entry, r); err != nil {
+	if _, err := io.Copy(f, r); err != nil {
+		return nil, err
+	}
+	if err := os.Rename(f.Name(), fullpath); err != nil {
```

Comment:
this is a change from another PR you have open right? not a huge fan of grouping this in with testing stuff, but if the tests fail without it then this is fine

Reply:
Yeah I'll drop it from the other one. At the time I thought the flakiness was pretty minor, but actually running the tests makes it pretty clear that the flakiness is bad.

Reply:
I thought about separating that out to it's own PR, but until tests are actually running we won't have much confidence in the PR itself...

Reply:
yeah that's kinda what I figured, just wanted to be sure